mirage/unikernel.ml

```diff
@@ -530,15 +541,44 @@ module Main (N : Mirage_net.S) (ASSETS : Mirage_kv.RO) = struct
                   let resp = H2.Response.create ~headers `Bad_request in
                   Reqd.respond_with_string reqd resp ""))
 
-    let handler resolver js_file metrics_cache =
+    let handler resolver js_file =
       {
         Alpn.error;
         request =
           (fun flow dst reqd protocol ->
-            request resolver metrics_cache flow dst reqd protocol js_file);
+            request resolver flow dst reqd protocol js_file);
       }
 
-    let update_blocklist resolver : unit Lwt.t =
+    let update_blocklist http_client resolver : unit Lwt.t =
+      let update_one source =
+        let ingest resp acc data =
+          if H2.Status.is_successful resp.Http_mirage_client.status then
+            Buffer.add_string acc data;
+          Lwt.return acc
+        in
+        Http_mirage_client.request http_client source ingest (Buffer.create 65536) >>= function
+        | Error e ->
+          Logs.warn (fun m -> m "Failed retrieving blocklist from %S: %a" source Mimic.pp_error e);
+          Lwt.return_unit
+        | Ok (resp, acc) ->
+          if H2.Status.is_successful resp.status then
+            let blocklist = Buffer.contents acc in
+            Logs.info (fun m -> m "Retrieved data:@ %s" blocklist);
+            let blocklist = Blocklist.blocklist_of_string blocklist in
+            let trie = Resolver.primary_data resolver in
+            let trie = List.fold_left Blocklist.add_dns_entries trie blocklist in
+            Resolver.update_primary_data resolver trie;
```

Comment:
the issue I see here is that if the blocklist removes items, we're not able to do this. So, I guess we've to either start from an empty trie, or prune the blocklist entries from the trie. The latter is my preference.

Reply:
Hmm yes maybe we need to maintain a shadow blocklist entries. I also worry if it could be a vector to remove dns entries which are not blocklist entries.

Reply:
I updated https://github.com/mirage/ocaml-dns/pull/380 to check for the "blocked" SOA nameserver.